### PR TITLE
torch::nn::LayerNorm

### PR DESCRIPTION
**Pull request by anjali411**

### CODE REVIEW

File: torch/csrc/api/src/nn/modules/normalization.cpp [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    weight = register_parameter("weight", torch::empty(torch::IntArrayRef(options.normalized_shape())));
+    bias = register_parameter("bias", torch::empty(torch::IntArrayRef(options.normalized_shape())));
+  }
+}
+
+void LayerNormImpl::pretty_print(std::ostream& stream) const {
+  stream << std::boolalpha
+         << "torch::nn::LayerNorm(normalized_shape=" << torch::IntArrayRef(options.normalized_shape())
+         << ", elementwise_affine=" << options.elementwise_affine() << ", eps=" << options.eps()
+         << ")";
```
Comment (yf225) [inline torch/csrc/api/src/nn/modules/normalization.cpp]: The Python version seems to print `LayerNorm((1, 2), eps=1e-05, elementwise_affine=True)` - to match the Python version we likely don't need to print `normalized_shape=`, and `eps` should show before `elementwise_affine`.

File: torch/csrc/api/include/torch/nn/modules/normalization.h [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+  Tensor weight;
+
+  /// The learned bias.
+  /// Initialized to zeros `elementwise_affine` option is set to `true` upon construction.
+  Tensor bias;
+
+  /// A `ModuleHolder` subclass for `LayerNormImpl`.
+  /// See the documentation for `LayerNormImpl` class to learn what methods it
+  /// provides, or the documentation for `ModuleHolder` to learn about PyTorch's
+  /// module storage semantics.
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/modules/normalization.h]: I think this comment should be above `TORCH_MODULE(LayerNorm);` instead.

File: torch/csrc/api/src/nn/modules/normalization.cpp [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+namespace torch {
+namespace nn {
+
+LayerNormImpl::LayerNormImpl(const LayerNormOptions& options_) : options(options_) {
+  reset();
+}
+
+void LayerNormImpl::reset() {
+  if (options.elementwise_affine()) {
+    weight = register_parameter("weight", torch::ones(torch::IntArrayRef(options.normalized_shape())));
```
Comment (yf225) [inline torch/csrc/api/src/nn/modules/normalization.cpp]: I think the following should work, because `std::vector<int64_t>` can be converted to `torch::IntArrayRef` implicitly:
```suggestion
    weight = register_parameter("weight", torch::ones(options.normalized_shape()));
```

File: torch/csrc/api/src/nn/modules/normalization.cpp [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    bias = register_parameter("bias", torch::zeros(torch::IntArrayRef(options.normalized_shape())));
```
Comment (yf225) [inline torch/csrc/api/src/nn/modules/normalization.cpp]: similarly here, we should be able to remove `torch::IntArrayRef`.

File: torch/csrc/api/src/nn/modules/normalization.cpp (hunk 3 above)
Comment (yf225) [reply torch/csrc/api/src/nn/modules/normalization.cpp]: Is there a reason we use `torch::ones`? I believe to match the Python version we should use `torch::empty`.

File: torch/csrc/api/src/nn/modules/normalization.cpp (hunk 4 above)
Comment (yf225) [reply torch/csrc/api/src/nn/modules/normalization.cpp]: similarly here, we should probably use `torch::empty`. And do `torch::nn::init::zeros_(bias)` at the end of `reset()`.

File: torch/csrc/api/src/nn/modules/normalization.cpp [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+  } else {
+    weight = register_parameter("weight", torch::empty(torch::IntArrayRef(options.normalized_shape())));
+    bias = register_parameter("bias", torch::empty(torch::IntArrayRef(options.normalized_shape())));
```
Comment (yf225) [inline torch/csrc/api/src/nn/modules/normalization.cpp]: This if-branch maps to Python version:
```python
self.register_parameter('weight', None)
self.register_parameter('bias', None)
```
to achieve parity, the corresponding C++ version should be:
```cpp
weight = register_parameter("weight", torch::Tensor(), /*requires_grad=*/false);
bias = register_parameter("bias", torch::Tensor(), /*requires_grad=*/false);
```

File: torch/csrc/api/src/nn/modules/normalization.cpp (hunk 3 above)
Comment (yf225) [reply torch/csrc/api/src/nn/modules/normalization.cpp]: ah I see that you are moving the logic of `reset_parameters` to here. I think to match the Python version better, we can keep the `reset_parameters` logic at the end of the `reset()` call, like how the Python version does it.

File: torch/csrc/api/src/nn/modules/normalization.cpp [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+  }
```
Comment (yf225) [inline torch/csrc/api/src/nn/modules/normalization.cpp]: We should add:
```cpp
if (options.elementwise_affine()) {
    torch::nn::init::ones_(weight);
    torch::nn::init::zeros_(bias);
}
```

File: torch/csrc/api/src/nn/modules/normalization.cpp [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+void LayerNormImpl::pretty_print(std::ostream& stream) const {
+  stream << std::boolalpha
+         << "torch::nn::LayerNorm(normalized_shape=" << torch::IntArrayRef(options.normalized_shape())
+         << ", elementwise_affine=" << options.elementwise_affine() << ", eps=" << options.eps()
+         << ")";
+}
+
+torch::Tensor LayerNormImpl::forward(const Tensor& input) {
+  return torch::layer_norm(input, torch::IntArrayRef(options.normalized_shape()), weight, bias, options.eps(),
+                          at::globalContext().userEnabledCuDNN());
```
Comment (yf225) [inline torch/csrc/api/src/nn/modules/normalization.cpp]: As part of the PR, we should probably implement `F::layer_norm` (i.e. the functional form of LayerNorm) as well :) There are many good examples in https://github.com/pytorch/pytorch/issues/25883, for example https://github.com/pytorch/pytorch/pull/27446/files. After we have `F::layer_norm`, this `forward` function should call `F::layer_norm`, to match Python version better.

File: torch/csrc/api/include/torch/nn/modules/normalization.h [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+};
+
+TORCH_MODULE(LayerNorm);
```
Comment (yf225) [inline torch/csrc/api/include/torch/nn/modules/normalization.h]: We need to add functionality tests for the `torch::nn::LayerNorm` module, similar to how other `torch::nn` layer parity PRs (in https://github.com/pytorch/pytorch/issues/25883) does it. For the simplest tests, we can do (in Python version):
```
        >>> input = torch.randn(20, 5, 10, 10)
        >>> # With Learnable Parameters
        >>> m = nn.LayerNorm(input.size()[1:])
        >>> # Without Learnable Parameters
        >>> m = nn.LayerNorm(input.size()[1:], elementwise_affine=False)
        >>> # Normalize over last two dimensions
        >>> m = nn.LayerNorm([10, 10])
        >>> # Normalize over last dimension of size 10
        >>> m = nn.LayerNorm(10)
        >>> # Activating the module
        >>> output = m(input)
```
and check that for each module configuration, the output is what we expect it to be.

File: torch/csrc/api/src/nn/modules/normalization.cpp (hunk 7 above)
Comment (yf225) [reply torch/csrc/api/src/nn/modules/normalization.cpp]: After we add `F::layer_norm`, we should add tests for it as well.

**Review by yf225 — COMMENTED** [review]
Thanks a lot for the great PR @anjali411 ! I left some comments.

File: test/cpp/api/modules.cpp [hunk 9]
```diff
@@ -1661,6 +1661,15 @@ TEST_F(ModulesTest, PrettyPrintBatchNorm) {
       "torch::nn::BatchNorm(features=4, eps=0.5, momentum=0.1, affine=false, stateful=true)");
 }
 
+TEST_F(ModulesTest, PrettyPrintLayerNorm) {
+  ASSERT_EQ(
+    c10::str(LayerNorm(LayerNormOptions(std::vector<int64_t>({2, 2})))),
```
[the remark this answers is not in the document]
Comment (anjali411) [reply test/cpp/api/modules.cpp]: that's a good point! just tried, and it works!

[the remark this answers is not in the document]
Comment (yf225) [reply test/cpp/api/modules.cpp]: that's awesome!

File: torch/csrc/api/src/nn/modules/normalization.cpp (hunk 5 above)
Comment (anjali411) [reply torch/csrc/api/src/nn/modules/normalization.cpp]: alright

File: test/cpp/api/functional.cpp [hunk 10]
```diff
@@ -618,7 +627,7 @@ TEST_F(FunctionalTest, Normalize) {
 
     ASSERT_TRUE(torch::allclose(output, expected));
   }
-
```
Comment (yf225) [inline test/cpp/api/functional.cpp]: Thanks for fixing the extra spaces :D

File: test/cpp/api/modules.cpp [hunk 11]
```diff
@@ -1661,6 +1661,15 @@ TEST_F(ModulesTest, PrettyPrintBatchNorm) {
...
 }
 
+TEST_F(ModulesTest, PrettyPrintLayerNorm) {
+  ASSERT_EQ(
+    c10::str(LayerNorm(LayerNormOptions({2, 2}))),
+      "torch::nn::LayerNorm([2, 2], eps=1e-05, elementwise_affine=true)");
+      ASSERT_EQ(
+        c10::str(LayerNorm(LayerNormOptions({2, 2}).elementwise_affine(false).eps(2e-5))),
+          "torch::nn::LayerNorm([2, 2], eps=2e-05, elementwise_affine=false)");
+}
```
Comment (yf225) [inline test/cpp/api/modules.cpp]: It would be awesome to add a test for `LayerNorm` module forward as well, similar to other modules in this test file

File: torch/csrc/api/src/nn/modules/normalization.cpp [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+
+namespace torch {
+namespace nn {
+
+LayerNormImpl::LayerNormImpl(const LayerNormOptions& options_) : options(options_) {
+  reset();
+  if (options.elementwise_affine()) {
+    torch::nn::init::ones_(weight);
+    torch::nn::init::zeros_(bias);
+  }
```
Comment (yf225) [inline torch/csrc/api/src/nn/modules/normalization.cpp]: This code block
```cpp
  if (options.elementwise_affine()) {
    torch::nn::init::ones_(weight);
    torch::nn::init::zeros_(bias);
  }
```
should be at the end the `reset()` function, because the expectation is that all initialization logic lives in the `reset()` function :D

**Comment by yf225** [comment]
@anjali411 This PR has some conflicts with master, and it also contains changes to `third_party/fbgemm` which should be removed. We can work on those two tomorrow items if you need any help from me :)

**Comment by yf225** [comment]
@anjali411 The code conflict and submodule issue should be fixed :D

File: torch/csrc/api/src/nn/modules/normalization.cpp (hunk 12 above)
Comment (anjali411) [reply torch/csrc/api/src/nn/modules/normalization.cpp]: alright

File: test/cpp/api/functional.cpp [hunk 13]
```diff
@@ -687,6 +687,15 @@ TEST_F(FunctionalTest, PReLU) {
   ASSERT_TRUE(torch::allclose(y, y_exp));
 }
 
+TEST_F(FunctionalTest, LayerNorm) {
+  const auto input = torch::randn({2, 2});
+  const auto weight = torch::randn({2, 2});
+  const auto bias = torch::randn({2, 2});
+  auto y = F::layer_norm(input, LayerNormOptions({2, 2}).eps(2e-5), weight, bias);
```
Comment (yf225) [inline test/cpp/api/functional.cpp]: In order to test the "no weights and no bias" case for `F::layer_norm`, let's probably change this to
```suggestion
  auto y = F::layer_norm(input, LayerNormOptions({2, 2}).eps(2e-5));
```
and remove `weight` and `bias` for this test.

File: test/cpp/api/functional.cpp [hunk 14]
```diff
@@ -687,6 +687,15 @@ TEST_F(FunctionalTest, PReLU) {
...
+  auto y_exp = torch::layer_norm(input, {2, 2}, weight, bias, 2e-5);
```
Comment (yf225) [inline test/cpp/api/functional.cpp]: and we can pass `Tensor()` as `weight` and `bias` here.

File: torch/csrc/api/src/nn/modules/normalization.cpp [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+
+#include <ostream>
+#include <utility>
+
+namespace F = torch::nn::functional;
+
+namespace torch {
+namespace nn {
+
+LayerNormImpl::LayerNormImpl(const LayerNormOptions& options_) : options(options_) {
```
Comment (yf225) [inline torch/csrc/api/src/nn/modules/normalization.cpp]: based on clang-tidy:
```
{
    path: 'torch/csrc/api/src/nn/modules/normalization.cpp',
    start_line: 15,
    end_line: 15,
    start_column: 30,
    end_column: 30,
    annotation_level: 'failure',
    message: '[[modernize-pass-by-value]] warning: pass by value and use std::move'
  },
```
we should probably add ` // NOLINT(modernize-pass-by-value)` at the end of this line.

File: torch/csrc/api/src/nn/options/normalization.cpp [hunk 16]
```diff
@@ -1 +1,9 @@
 #include <torch/nn/options/normalization.h>
+
+namespace torch {
+namespace nn {
+
+LayerNormOptions::LayerNormOptions(std::vector<int64_t> normalized_shape) : normalized_shape_(normalized_shape) {}
```
Comment (yf225) [inline torch/csrc/api/src/nn/options/normalization.cpp]: clang-tidy complains:
```
{
    path: 'torch/csrc/api/src/nn/options/normalization.cpp',
    start_line: 6,
    end_line: 6,
    start_column: 36,
    end_column: 36,
    annotation_level: 'failure',
    message: '[[modernize-pass-by-value]] warning: pass by value and use std::move'
  }
```
for this one I think we can do `normalized_shape_(std::move(normalized_shape))`.

**Review by yf225 — COMMENTED** [review]
@anjali411 Thanks a lot for the update! I think we are almost there :D

**Review by yf225 — APPROVED** [review]
